File: tapyrus_wallet_source/rust-tapyrus-wallet-ffi-0.1.2/TapyrusWalletAndroid/example/app/src/main/java/com/chaintope/tapyrus/wallet/example/TapyrusWalletManager.kt
package com.chaintope.tapyrus.wallet.example

import android.content.Context
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.chaintope.tapyrus.wallet.Config
import com.chaintope.tapyrus.wallet.HdWallet
import com.chaintope.tapyrus.wallet.Network
import com.chaintope.tapyrus.wallet.TransferParams
import com.chaintope.tapyrus.wallet.generateMasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Manager class for Tapyrus Wallet operations
 */
class TapyrusWalletManager(val context: Context) {
    companion object {
        private const val TAG = "TapyrusWalletManager"
        private const val DB_FILENAME = "tapyrus_wallet.db"
        private const val MASTER_KEY_PREF = "tapyrus_master_key"
        private const val PREF_NAME = "tapyrus_wallet_prefs"
    }

    // State properties
    var currentAddress by mutableStateOf("")
        private set
    var balance by mutableStateOf(0.0)
        private set
    var isSyncing by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    // Wallet instances
    private var config: Config? = null
    private var wallet: HdWallet? = null
    private var isInitialized = false

    /**
     * Initialize the wallet
     */
    suspend fun initialize() {
        if (isInitialized) return
        
        withContext(Dispatchers.IO) {
            try {
                // Ensure JNA is loaded
                JnaLoader.load(context)
                
                // Get or create master key
                val masterKey = getMasterKey()

                // Get database file path
                val dbFilePath = File(context.filesDir, DB_FILENAME).absolutePath
                
                // Create config
                val networkMode = Network.PROD
                config = Config(
                    networkMode = networkMode,
                    networkId = 1939510133u,
                    genesisHash = "038b114875c2f78f5a2fd7d8549a905f38ea5faee6e29a3d79e547151d6bdd8a",
                    esploraUrl = "http://10.0.2.2:3001", // 10.0.2.2 is the host for the emulator
                    masterKey = masterKey,
                    dbFilePath = dbFilePath
                )
                
                // Create wallet
                wallet = HdWallet(config!!)
                
                // Mark as initialized before calling methods that might check this flag
                isInitialized = true
                
                // Get initial address if none exists
                if (currentAddress.isEmpty()) {
                    getNewAddressInternal()
                }
                
                // Sync wallet to get latest balance
                isSyncing = true
                wallet?.sync()
                isSyncing = false
                
                // Update balance
                updateBalanceInternal()
                errorMessage = null
            } catch (e: Exception) {
                Log.e(TAG, "Failed to initialize wallet: ${e.message}", e)
                errorMessage = "Failed to initialize wallet: ${e.message}"
            }
        }
    }

    /**
     * Get the master key from preferences or generate a new one
     */
    private fun getMasterKey(): String {
        val prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        var masterKey = prefs.getString(MASTER_KEY_PREF, null)
        
        if (masterKey == null) {
            // Generate a new master key
            masterKey = generateMasterKey(Network.PROD)
            
            // Save to preferences
            prefs.edit().putString(MASTER_KEY_PREF, masterKey).apply()
            Log.d(TAG, "Generated and saved new master key")
        } else {
            Log.d(TAG, "Using existing master key from preferences")
        }
        
        return masterKey
    }

    /**
     * Generate a new address - internal implementation that doesn't check initialization
     */
    private suspend fun getNewAddressInternal(): String {
        return withContext(Dispatchers.IO) {
            try {
                val result = wallet?.getNewAddress(null)
                if (result != null) {
                    currentAddress = result.address
                    Log.d(TAG, "Generated new address: ${result.address}")
                    currentAddress
                } else {
                    throw Exception("Failed to generate address")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error generating address: ${e.message}", e)
                errorMessage = "Error generating address: ${e.message}"
                ""
            }
        }
    }
    
    /**
     * Generate a new address - public method that ensures initialization
     */
    suspend fun getNewAddress(): String {
        return withContext(Dispatchers.IO) {
            try {
                if (!isInitialized) {
                    initialize()
                    return@withContext currentAddress // Return the address created during initialization
                }
                
                getNewAddressInternal()
            } catch (e: Exception) {
                Log.e(TAG, "Error generating address: ${e.message}", e)
                errorMessage = "Error generating address: ${e.message}"
                ""
            }
        }
    }

    /**
     * Update the wallet balance - internal implementation that doesn't check initialization
     */
    private suspend fun updateBalanceInternal() {
        withContext(Dispatchers.IO) {
            try {
                wallet?.let {
                    // Get balance in satoshis and convert to TPC
                    val balanceInSatoshis = it.balance(null)
                    balance = balanceInSatoshis.toDouble() / 100_000_000.0
                    Log.d(TAG, "Updated balance: $balance TPC")
                }
                errorMessage = null
            } catch (e: Exception) {
                Log.e(TAG, "Error updating balance: ${e.message}", e)
                errorMessage = "Error updating balance: ${e.message}"
            }
        }
    }
    
    /**
     * Update the wallet balance - public method that ensures initialization
     */
    suspend fun updateBalance() {
        withContext(Dispatchers.IO) {
            try {
                if (!isInitialized) {
                    initialize()
                    return@withContext // Balance is already updated during initialization
                }
                
                updateBalanceInternal()
            } catch (e: Exception) {
                Log.e(TAG, "Error updating balance: ${e.message}", e)
                errorMessage = "Error updating balance: ${e.message}"
            }
        }
    }

    /**
     * Sync the wallet with the blockchain
     */
    suspend fun syncWallet() {
        withContext(Dispatchers.IO) {
            try {
                if (!isInitialized) {
                    initialize()
                    return@withContext // No need to sync immediately after initialization
                }
                
                isSyncing = true
                wallet?.sync()
                updateBalanceInternal() // Use internal method to avoid potential recursion
                isSyncing = false
                errorMessage = null
            } catch (e: Exception) {
                Log.e(TAG, "Error syncing wallet: ${e.message}", e)
                errorMessage = "Error syncing wallet: ${e.message}"
                isSyncing = false
            }
        }
    }

    /**
     * Transfer TPC to another address
     */
    suspend fun transfer(toAddress: String, amount: Double): String {
        return withContext(Dispatchers.IO) {
            try {
                if (!isInitialized) {
                    initialize()
                }
                
                // Convert amount from TPC to satoshis
                val amountInSatoshis = (amount * 100_000_000.0).toULong()
                
                // Create transfer parameters
                val transferParams = TransferParams(amount = amountInSatoshis, toAddress = toAddress)
                
                // Execute the transfer
                val txid = wallet?.transfer(params = listOf(transferParams), utxos = listOf()) ?: 
                    throw Exception("Wallet is not initialized")
                
                // Update balance after transfer
                updateBalanceInternal()
                
                errorMessage = null
                txid
            } catch (e: Exception) {
                Log.e(TAG, "Error transferring funds: ${e.message}", e)
                errorMessage = "Error transferring funds: ${e.message}"
                throw e
            }
        }
    }

    /**
     * Clean up resources
     */
    fun cleanup() {
        try {
            wallet?.close()
            config?.close()
            isInitialized = false
        } catch (e: Exception) {
            Log.e(TAG, "Error cleaning up wallet resources: ${e.message}", e)
        }
    }
}

File: tapyrus_wallet_source/rust-tapyrus-wallet-ffi-0.1.2/TapyrusWalletSwift/example/example/TapyrusWalletManager.swift
//
//  TapyrusWalletManager.swift
//  example
//
//  Created on 2025/02/26.
//

import Foundation
import Security
import TapyrusWallet

// Enum for wallet-related errors
enum WalletError: Error {
    case keyGenerationFailed
    case keychainSaveFailed
    case keychainLoadFailed
    case walletInitializationFailed
    case walletSyncFailed
    case addressGenerationFailed
}

// Class to manage HdWallet operations
class TapyrusWalletManager: ObservableObject {
    // Published properties for UI updates
    @Published var currentAddress: String = ""
    @Published var balance: Double = 0.0
    @Published var isSyncing: Bool = false
    
    // The wallet instance
    private var wallet: HdWallet?
    
    // Network mode
    private let networkMode: Network = .prod
    
    // Keychain constants
    private let keychainService = "com.example.TapyrusWalletApp"
    private let masterKeyAccount = "masterKey"
    
    // Initialize the wallet manager
    init() {
        do {
            try setupWallet()
        } catch {
            print("Failed to setup wallet: \(error)")
        }
    }
    
    // Setup the wallet - create or load from storage
    private func setupWallet() throws {
        // Check if we have a master key in the keychain
        var masterKey: String
        
        if let storedMasterKey = loadMasterKeyFromKeychain() {
            // Use the stored master key
            masterKey = storedMasterKey
            print("Using existing master key from keychain")
        } else {
            // Generate a new master key
            masterKey = TapyrusWallet.generateMasterKey(networkMode: networkMode)
            
            // Save the master key to keychain
            if !saveMasterKeyToKeychain(masterKey) {
                throw WalletError.keychainSaveFailed
            }
            print("Generated and saved new master key")
        }
        
        // Get the Documents directory path which is more accessible in iOS apps
        let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dbFilePath = documentsDir.appendingPathComponent("tapyrus_wallet.db").path
        
        // Ensure the directory exists
        try? FileManager.default.createDirectory(at: documentsDir, withIntermediateDirectories: true)
        
        print("Database path: \(dbFilePath)")
        
        // Create wallet configuration with testnet values and the master key
        let config = Config(
            networkMode: networkMode,
            networkId: 1939510133,
            genesisHash: "038b114875c2f78f5a2fd7d8549a905f38ea5faee6e29a3d79e547151d6bdd8a",
            esploraUrl: "http://localhost:3001",
            masterKey: masterKey,
            dbFilePath: dbFilePath
        )
        
        // Create a new wallet with the configuration
        do {
            wallet = try HdWallet(config: config)
            
            // Perform initial sync
            syncWallet()
        } catch {
            print("Error initializing wallet: \(error)")
            throw WalletError.walletInitializationFailed
        }
    }
    
    // Save master key to keychain
    private func saveMasterKeyToKeychain(_ key: String) -> Bool {
        guard let data = key.data(using: .utf8) else {
            return false
        }
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: masterKeyAccount,
            kSecValueData as String: data
        ]
        
        // Delete any existing key before saving
        SecItemDelete(query as CFDictionary)
        
        // Add the new key
        let status = SecItemAdd(query as CFDictionary, nil)
        return status == errSecSuccess
    }
    
    // Load master key from keychain
    private func loadMasterKeyFromKeychain() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: masterKeyAccount,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess, let data = result as? Data, let masterKey = String(data: data, encoding: .utf8) {
            return masterKey
        }
        return nil
    }
    
    // Sync the wallet with the blockchain
    func syncWallet() {
        guard let wallet = wallet else { return }
        
        DispatchQueue.global(qos: .background).async { [weak self] in
            self?.isSyncing = true
            
            do {
                // Perform full sync
                try wallet.fullSync()
                
                // Update balance
                self?.updateBalance()
                
                DispatchQueue.main.async {
                    self?.isSyncing = false
                }
            } catch {
                print("Sync error: \(error)")
                DispatchQueue.main.async {
                    self?.isSyncing = false
                }
            }
        }
    }
    
    // Generate a new address
    func getNewAddress() -> String {
        guard let wallet = wallet else { return "" }
        debugPrint(wallet)
        
        do {
            // Call the wallet's getNewAddress function with colorId set to nil
            let result = try wallet.getNewAddress(colorId: nil)
            let address = result.address
            
            DispatchQueue.main.async { [weak self] in
                self?.currentAddress = address
            }
            
            return address
        } catch {
            print("Error generating address: \(error)")
            return ""
        }
    }
    
    // Update the wallet balance
    private func updateBalance() {
        guard let wallet = wallet else { return }
        
        do {
            // Call the wallet's balance function with colorId set to nil for TPC
            let balanceValue = try wallet.balance(colorId: nil)
            
            // Convert to Double for display (balance is in satoshis, convert to TPC)
            let tpcBalance = Double(balanceValue) / 100_000_000.0
            
            DispatchQueue.main.async { [weak self] in
                self?.balance = tpcBalance
            }
        } catch {
            print("Error getting balance: \(error)")
        }
    }
    
    // Transfer TPC to another address
    func transfer(toAddress: String, amount: Double) async throws -> String {
        guard let wallet = wallet else {
            throw WalletError.walletInitializationFailed
        }
        
        // Convert amount from TPC to satoshis (smallest unit)
        let amountInSatoshis = UInt64(amount * 100_000_000.0)
        
        // Create transfer parameters
        let transferParams = TransferParams(amount: amountInSatoshis, toAddress: toAddress)
        
        // Execute the transfer
        let txid = try wallet.transfer(params: [transferParams], utxos: [])
        
        // Update balance after transfer
        updateBalance()
        
        return txid
    }
}
